File: exam_prep/IST603/notebook_app/note.py
"""
Note class per UML: stores memo, creation_date, tags; provides match(search_filter).
"""

from datetime import datetime


class Note:
    """Individual note with memo, creation date, and tags."""

    def __init__(self, note_id: int, memo: str, tags: str = "") -> None:
        self.memo = memo
        self.creation_date = datetime.now()
        self.tags = tags
        self._note_id = note_id  # id set by Notebook; kept for modify_memo/modify_tags

    @property
    def note_id(self) -> int:
        return self._note_id

    def match(self, search_filter: str) -> bool:
        """Return True if this note matches the search filter (in memo or tags)."""
        if not search_filter:
            return True
        q = search_filter.lower()
        return q in self.memo.lower() or q in self.tags.lower()

File: exam_prep/IST603/notebook_app/java/ist603/notebook/Note.java
package ist603.notebook;

import java.time.LocalDateTime;

/**
 * Note per UML: stores memo, creation_date, tags; provides match(search_filter).
 */
public class Note {
    public String memo;
    public LocalDateTime creationDate;
    public String tags;
    private final int noteId;

    public Note(int noteId, String memo, String tags) {
        this.noteId = noteId;
        this.memo = memo;
        this.creationDate = LocalDateTime.now();
        this.tags = tags != null ? tags : "";
    }

    public int getNoteId() {
        return noteId;
    }

    /**
     * Returns true if this note matches the search filter (in memo or tags).
     */
    public boolean match(String searchFilter) {
        if (searchFilter == null || searchFilter.isEmpty()) {
            return true;
        }
        String q = searchFilter.toLowerCase();
        return (memo != null && memo.toLowerCase().contains(q))
                || (tags != null && tags.toLowerCase().contains(q));
    }
}
